src/centroiders.cpp

```diff
@@ -1,325 +1,558 @@
-        sq_totalMag += image[i] * image[i];
-    }
-    float mean = totalMag / totalPixels;
-    float variance = (sq_totalMag / totalPixels) - (mean * mean);
-    std = std::sqrt(variance);
-    return mean + (std * 5);
-}
+// Generic functor
+template <typename _Scalar, int NX = Eigen::Dynamic, int NY = Eigen::Dynamic>
+struct Functor {
+  typedef _Scalar Scalar;
+  enum { InputsAtCompileTime = NX, ValuesAtCompileTime = NY };
+  typedef Eigen::Matrix<Scalar, InputsAtCompileTime, 1> InputType;
+  typedef Eigen::Matrix<Scalar, ValuesAtCompileTime, 1> ValueType;
+  typedef Eigen::Matrix<Scalar, ValuesAtCompileTime, InputsAtCompileTime> JacobianType;
 
-struct CentroidParams {
-    float yCoordMagSum;
-    float xCoordMagSum;
-    long magSum;
-    int xMin;
-    int xMax;
-    int yMin;
-    int yMax;
-    int cutoff;
-    bool isValid;
-    std::unordered_set<int> checkedIndices;
+  int m_inputs, m_values;
+
+  Functor() : m_inputs(InputsAtCompileTime), m_values(ValuesAtCompileTime) {}
+  Functor(int inputs, int values) : m_inputs(inputs), m_values(values) {}
+
+  int inputs() const { return m_inputs; }
+  int values() const { return m_values; }
 };
 
-//recursive helper here
-void CogHelper(CentroidParams *p, long i, unsigned char *image, int imageWidth, int imageHeight) {
 
-    if (i >= 0 && i < imageWidth * imageHeight && image[i] >= p->cutoff && p->checkedIndices.count(i) == 0) {
-        //check if pixel is on the edge of the image, if it is, we dont want to centroid this star
-        if (i % imageWidth == 0 || i % imageWidth == imageWidth - 1 || i / imageWidth == 0 || i / imageWidth == imageHeight - 1) {
-            p->isValid = false;
-        }
-        p->checkedIndices.insert(i);
-        if (i % imageWidth > p->xMax) {
-            p->xMax = i % imageWidth;
-        } else if (i % imageWidth < p->xMin) {
-            p->xMin = i % imageWidth;
-        }
-        if (i / imageWidth > p->yMax) {
-            p->yMax = i / imageWidth;
-        } else if (i / imageWidth < p->yMin) {
-            p->yMin = i / imageWidth;
-        }
-        p->magSum += image[i];
-        p->xCoordMagSum += ((i % imageWidth)) * image[i];
-        p->yCoordMagSum += ((i / imageWidth)) * image[i];
-        if (i % imageWidth != imageWidth - 1) {
-            CogHelper(p, i + 1, image, imageWidth, imageHeight);
-        }
-        if (i % imageWidth != 0) {
-            CogHelper(p, i - 1, image, imageWidth, imageHeight);
-        }
-        CogHelper(p, i + imageWidth, image, imageWidth, imageHeight);
-        CogHelper(p, i - imageWidth, image, imageWidth, imageHeight);
+struct LSGFFunctor : Functor<double> {
+  // First param = number of parameters in your model
+  // Second param = number of data points you want to test over
+  // for us, = 2 * nb + 1
+  LSGFFunctor(const int nb, const int marg, Eigen::VectorXd X, const unsigned char *image,
+              const int w, const int x0, const int y0)
+      : Functor<double>(3, 2 * nb + 1),
```

Comment:
3 = number of parameters we're trying to find (a, xb or yb, sigma), 2 * nb + 1 = number of data points for least-squares